crates/runtime/src/opentelemetry.rs

```diff
@@ -69,8 +132,231 @@ impl MetricsService for Service {
     }
 }
 
-pub async fn start(bind_address: SocketAddr) -> Result<()> {
-    let service = Service;
+pub fn metric_data_to_record_batch(metric: &str, data: &Data) -> (Result<RecordBatch>, u64) {
+    match data {
+        Data::Gauge(gauge) => (
+            number_data_points_to_record_batch(metric, &gauge.data_points),
+            gauge.data_points.len() as u64,
+        ),
+        Data::Sum(sum) => (
+            number_data_points_to_record_batch(metric, &sum.data_points),
+            sum.data_points.len() as u64,
+        ),
+        _ => (UnsupportedMetricDataTypeSnafu.fail(), 0),
+    }
+}
+
+fn number_data_points_to_record_batch(
+    metric: &str,
+    data_points: &Vec<NumberDataPoint>,
+) -> Result<RecordBatch> {
+    let mut values_builder = Float64Builder::new();
+    let mut attributes = Vec::new();
+    for data_point in data_points {
+        if let Some(value) = &data_point.value {
+            match value {
+                Value::AsDouble(double_value) => {
+                    values_builder.append_value(*double_value);
+                }
+                Value::AsInt(int_value) => {
+                    values_builder.append_value(*int_value as f64);
+                }
+            }
+        } else {
+            values_builder.append_null();
+        }
+        attributes.push(&data_point.attributes);
+    }
+
+    let mut fields = vec![Field::new("value", DataType::Float64, true)];
+    let mut columns: Vec<ArrayRef> = vec![Arc::new(values_builder.finish())];
+
+    let (attribute_fields_map, attribute_columns_map) =
+        attributes_to_fields_and_columns(metric, attributes);
+    let attribute_fields: Vec<Field> = attribute_fields_map.into_iter().map(|(_, v)| v).collect();
+    fields.extend(attribute_fields);
+    columns.extend(
+        attribute_columns_map
+            .into_iter()
+            .map(|(_, mut v)| v.finish()),
+    );
+
+    match RecordBatch::try_new(Arc::new(Schema::new(fields)), columns) {
+        Ok(record_batch) => Ok(record_batch),
+        Err(e) => Err(e).context(FailedToBuildRecordBatchSnafu),
+    }
+}
+
+#[allow(clippy::too_many_lines)]
+fn attributes_to_fields_and_columns(
+    metric: &str,
+    attributes: Vec<&Vec<KeyValue>>,
+) -> (
+    IndexMap<String, Field>,
+    IndexMap<String, Box<dyn ArrayBuilder>>,
+) {
+    let mut fields: IndexMap<String, Field> = IndexMap::new();
+    let mut columns: IndexMap<String, Box<dyn ArrayBuilder>> = IndexMap::new();
+    for inner_attributes in attributes {
+        for attribute in inner_attributes {
+            let key_str = attribute.key.as_str();
+            match &attribute.value {
+                Some(any_value) => match &any_value.value {
+                    Some(value) => match value {
+                        any_value::Value::StringValue(string_value) => {
+                            match columns.get_mut(key_str) {
```

Comment:
Unfortunately these branches can't be generalized because `new` and `append` are not defined by a trait, they are defined by each builder individually.

Reply:
Per @y-f-u I'm going to look into using macros for this